### PR TITLE
Merge20211111

### PR DESCRIPTION
**Pull request by samtertzakian**
Merge20211111
1. Update DMF_UdeClient to correct issues with Manual dispatch type endpoints.
2. Correct bug in DMF_DeviceInterfaceTarget when device interface is disabled. Also, update unit tests to test disable/enable of device interface targets.
3. Add DMF_GpioTarget and DMF_SpiTarget to User-mode Library (they are already in Kernel-mode Library).
4. DMF_USER_MODE is now automatically defined based on UMDF settings. Thus, it is no longer necessary for Client User-mode drivers to define DMF_USER_MODE as it was in the past.

### CODE REVIEW

File: Dmf/Modules.Library/Dmf_UdeClient.c [hunk 1]
```diff
@@ -1691,6 +1821,105 @@ Return Value:
...
+/*++
+
+Routine Description:
+
+    Get the address from a given endpoint.
+
+Arguments:
+
+    DmfModule - This Module's handle.
+    Endpoint - The given endpoint.
```
Comment (ankurung) [inline Dmf/Modules.Library/Dmf_UdeClient.c:1842]: an argument is missing here.

File: Dmf/Modules.Library/Dmf_UdeClient.md [hunk 2]
```diff
@@ -403,10 +463,67 @@ NTSTATUS
...
+##### Returns
+
+None
+
+##### Parameters
+Parameter | Description
+----|----
+DmfModule | An open DMF_UdeClient Module handle.
+Endpoint | The given endpoint.
+UdecxUsbDevice | The device to which the endpoint is attached.
```
Comment (ankurung) [inline Dmf/Modules.Library/Dmf_UdeClient.md:496]: located?

File: Dmf/Modules.Library/Dmf_UdeClient.c (hunk 1 above)
Comment (samtertzakian) [reply Dmf/Modules.Library/Dmf_UdeClient.c:1842]: i don't see anything missing...

````
_IRQL_requires_max_(PASSIVE_LEVEL)
VOID
DMF_UdeClient_DeviceEndpointInformationGet(
    _In_ DMFMODULE DmfModule,
    _In_ UDECXUSBENDPOINT Endpoint,
    _Out_opt_ UDECXUSBDEVICE* UdecxUsbDevice,
    _Out_opt_ UCHAR* Address
    )
````

Gets information about a given endpoint.

##### Returns

None

##### Parameters
Parameter | Description
----|----
DmfModule | An open DMF_UdeClient Module handle.
Endpoint | The given endpoint.
UdecxUsbDevice | The device to which the endpoint is attached.
Address | The address of the endpoint.

File: Dmf/Modules.Library/Dmf_UdeClient.md (hunk 2 above)
Comment (samtertzakian) [reply Dmf/Modules.Library/Dmf_UdeClient.md:496]: I don't see anything missing...

````
_IRQL_requires_max_(PASSIVE_LEVEL)
VOID
DMF_UdeClient_DeviceEndpointInformationGet(
    _In_ DMFMODULE DmfModule,
    _In_ UDECXUSBENDPOINT Endpoint,
    _Out_opt_ UDECXUSBDEVICE* UdecxUsbDevice,
    _Out_opt_ UCHAR* Address
    )
````

Gets information about a given endpoint.

##### Returns

None

##### Parameters
Parameter | Description
----|----
DmfModule | An open DMF_UdeClient Module handle.
Endpoint | The given endpoint.
UdecxUsbDevice | The device to which the endpoint is attached.
Address | The address of the endpoint.

File: Dmf/Modules.Library/Dmf_UdeClient.c (hunk 1 above)
Comment (samtertzakian) [reply Dmf/Modules.Library/Dmf_UdeClient.c:1842]: fixed

File: Dmf/Modules.Library/Dmf_UdeClient.md (hunk 2 above)
Comment (samtertzakian) [reply Dmf/Modules.Library/Dmf_UdeClient.md]: i understand now...fixed

**Review by ankurung — APPROVED** [review]

**Pull request merged**